### PR TITLE
change API to allow for multiple and potentially different server runtimes

### PR DESCRIPTION
**Pull request by dario-piotrowicz**

### CODE REVIEW

File: packages/shared-vite-runtime-utils/src/index.ts [hunk 1]
```diff
@@ -1,17 +1,27 @@
...
+    createServerRuntime: CreateServerRuntime;
   }
 }
 
-export type SSRRuntime = {
+type ServerRuntimeFactory = (
+  runtimeOptions?: Record<string, unknown>,
+) => Promise<ServerRuntime>;
+
+type CreateServerRuntime = (
```
Comment (IgorMinar) [inline packages/shared-vite-runtime-utils/src/index.ts]: q: I wonder why you didn't inline this type on line #11. I personally find this explicit type definition not very useful and it makes reading the code harder because I need to jump within a file to know what's going on. The code you have is not wrong, it's just more complex than it needs to be.

btw the `ServerRuntimeFactory` type above is referred to in other parts of this PR so it makes sense that you externalized it.

File: examples/basic-handler/vite.config.js [hunk 2]
```diff
@@ -42,7 +54,10 @@ export default {
   optimizeDeps: {
     include: [],
   },
-  plugins: [runtime(), basicHandler({ entrypoint: './entry.ts' })],
+  plugins: [
+    ...runtimeRegistrationPlugins,
+    basicHandler({ entrypoint: './entry.ts', serverRuntimeId: chosenRuntime }),
```
Comment (IgorMinar) [inline examples/basic-handler/vite.config.js]:
```suggestion
    // note that the application developer can via vite.config.js decide which server runtime to use
    // the server runtime choice belongs to the developer
    basicHandler({ entrypoint: './entry.ts', serverRuntimeId: chosenRuntime }),
```

File: packages/vite-runtime-node-plugin/src/index.ts [hunk 3]
```diff
@@ -7,34 +7,45 @@ import { type ViteDevServer } from 'vite';
...
-        ssrRuntimeResolve = resolve;
-      });
-      server.ssrRuntime$ = ssrRuntime$;
+    configureServer(server: ViteDevServer) {
+      // IMPORTANT: the following line should not be needed
+      //            if ViteDevServer were to have the runtime
+      //            registration built in
+      setupServerRuntimeRegistration(server);
+
+      const runtimeCreation = () => {
```
Comment (IgorMinar) [inline packages/vite-runtime-node-plugin/src/index.ts]: Similar to the comment earlier, consider inlining this function into line 48 below to improve readability.

File: packages/vite-runtime-workerd-plugin/src/index.ts [hunk 4]
```diff
@@ -4,34 +4,45 @@ import { Miniflare } from 'miniflare';
...
+      const runtimeCreation = () => {
+        let runtimeResolve: (runtime: ServerRuntime) => void;
+        const runtimePromise = new Promise<ServerRuntime>(resolve => {
+          runtimeResolve = resolve;
+        });
 
-      server.httpServer.once('listening', () => {
-        setupFetchModuleEndpoint(server);
+        server.httpServer.once('listening', () => {
+          setupFetchModuleEndpoint(server);
```
Comment (IgorMinar) [inline packages/vite-runtime-workerd-plugin/src/index.ts]: observation: since this is a shared helper function, we'll be setting up the endpoint once per server runtime that uses the helper function, but because they override each other, only the first registration will be effective. Ideally the module fetching endpoint is built into Vite proper, but if that's not going to happen, we should modify the implementation of `setupFetchModuleEndpoint` to register the middleware endpoint only once.

File: examples/basic-handler/vite.config.js [hunk 5]
```diff
@@ -1,13 +1,18 @@
...
+export function basicHandler({ entrypoint, serverRuntimeId }) {
   return {
     name: 'basic-handler-plugin',
+    /** @param {import('vite').ViteDevServer} server */
     configureServer(server) {
       return async () => {
-        const ssrRuntime = await server.ssrRuntime$;
-        const dispatchRequest = await ssrRuntime.createRequestDispatcher({
+        // Note: here we create an instance of the chosenRuntime, interestingly we can:
+        //       - create more instances of the same runtime (with maybe different options for example)
```
Comment (IgorMinar) [inline examples/basic-handler/vite.config.js]: I'm realizing only now as I read this, that in order to create a runtime with different options, we'd need refactor the `basicHandler` function to accept an instance of ServerRuntime rather than (or in addition to) the id. That way the app developer can configure the server runtime to their hearts content and pass it in, or just use the runtime name and get a default instance.

let's discuss this a bit more when we talk tomorrow.

**Review by IgorMinar — COMMENTED** [review]
Thanks for putting this together @dario-piotrowicz, I learned a lot when reading the implementation and left a few comments.

This think that this change is heading in the right direction which enables multiple instances of the same or different server runtimes to be initialized, but it also raises some questions as to how this should work in practice.

I think it would be the best to bring this feature up for a discussion with the Vite team and framework teams that are interested in this feature (e.g. Astro) before we continue iterating on this.

btw, we'll also need to take into account Patak's https://github.com/vitejs/vite/pull/16003 which is very related to supporting multiple server runtimes.

File: packages/shared-vite-runtime-utils/src/index.ts (hunk 1 above)
Comment (dario-piotrowicz) [reply packages/shared-vite-runtime-utils/src/index.ts]: I guess it's just personal preference 😅

Inlining types might lead to bigger less readable types in my opinion, so separating things and giving them meaningful name helps making things more clear I think

anyways in this case this type name is not that much clearer than the inlined version, and also I did inline the `registerServerRuntime` one so I might as well inline this as well 😅

TLDR; yeah this one is my bad sorry 😅, but generally externalizing types is, for me, like using auxiliary/informatory variables, in the sense that you add extra (unnecessary) code for the sake of:
 - making the base type simpler to read at a first glance
 - getting an extra identifier that you can use to convey meaning

File: packages/vite-runtime-workerd-plugin/src/index.ts (hunk 4 above)
Comment (dario-piotrowicz) [reply packages/vite-runtime-workerd-plugin/src/index.ts]: good point... 🤔 :+1:

File: examples/basic-handler/vite.config.js (hunk 5 above)
Comment (dario-piotrowicz) [reply examples/basic-handler/vite.config.js]: yes exactly, I made the options optional and in this simple example the runtimes we define just happen not to specify and option, but they could

File: packages/shared-vite-runtime-utils/src/index.ts (hunk 1 above)
Comment (dario-piotrowicz) [reply packages/shared-vite-runtime-utils/src/index.ts]: Update 🙂

File: examples/basic-handler/vite.config.js (hunk 5 above)
Comment (dario-piotrowicz) [reply examples/basic-handler/vite.config.js]: I've updated this to allow and showcase how to for example tweak the `inspectorPort` of Miniflare

Hopefully this is along the lines of what you were thinking? 🙂

File: packages/vite-runtime-node-plugin/src/index.ts (hunk 3 above)
Comment (dario-piotrowicz) [reply packages/vite-runtime-node-plugin/src/index.ts]: Updated 🙂👍

File: packages/vite-runtime-workerd-plugin/src/index.ts (hunk 4 above)
Comment (dario-piotrowicz) [reply packages/vite-runtime-workerd-plugin/src/index.ts]: I've added a simple check in `setupFetchModuleEndpoint` to make sure that we don't register the endpoint more than once 🙂👍
